src/ray/core_worker/core_worker.cc

```diff
@@ -190,13 +194,22 @@ CoreWorker::CoreWorker(const WorkerType worker_type, const Language language,
 
   task_manager_.reset(new TaskManager(
       memory_store_, reference_counter_, actor_manager_,
-      [this](const TaskSpecification &spec) {
+      /*retry_task_callback=*/[this](const TaskSpecification &spec) {
         // Retry after a delay to emulate the existing Raylet reconstruction
         // behaviour. TODO(ekl) backoff exponentially.
         RAY_LOG(ERROR) << "Will resubmit task after a 5 second delay: "
                        << spec.DebugString();
+        const auto now = current_time_ms();
         absl::MutexLock lock(&mutex_);
-        to_resubmit_.push_back(std::make_pair(current_time_ms() + 5000, spec));
+        if (spec.IsActorTask()) {
+          const auto actor_id = spec.ActorId();
+          // Schedule a kill task unless we already have one in flight.
+          if (!pending_kill_for_actor_.contains(actor_id)) {
+            pending_actor_kills_.emplace_back(now + kActorRestartDeadlineMillis, spec.ActorId());
```

Comment:
If I understand this correctly, the idea is to kill an actor whose tasks have failed if we don't get a notification about it being reconstructed within the deadline (currently hardcoded to 3s). Is this to prevent against tasks that fail without crashing the actor?

I'm not sure that we need this at all - can't we just delegate this handling to the raylet (for now, eventually GCS service/owner) and then here wait for the state to transition via GCS notifications? AFAIK the raylet currently guarantees that the actor's state will eventually transition to the correct state (unless the raylet crashes, in which case its actors will be marked dead).

Reply:
Yeah I think we don't need to keep killing the actors anymore. It was important in my previous patch, but now I can just keep resubmitting the tasks indefinitely.